torch/_dynamo/side_effects.py

```diff
@@ -375,6 +380,73 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 ]
             )
 
+    def register_hook(self, tensor, hook, handle):
+        self.tensor_hooks.append((tensor, hook, handle))
+
+    def codegen_hooks(self, cg):
+        for (
+            tensor,
+            hook,
+            handle,
+        ) in self.tensor_hooks:
+            # Note: [On tensor.register_hook]
+            #
+            # register_hook on a tensor, AKA backward hooks, have slightly nuanced differences in how they are implemented
+            # when it comes to hooks on objects with sources (inputs, params) vs objects without sources (intermediaries).
+            #
+            # For tensors with a source, we bypass direct inclusion of register_hook calls in the graph.
+            # Instead, these are tracked and stashed as a global variable, enabling their association with tensors in
+            # the residuals. During dynamo's frame creation, these hooks are invoked seamlessly on known reconstructible/fetch-able
+            # tensors. Because a source indicates knowledge of this object outside the torch compile region, and
+            # because we are running residuals firmly before .backward() can be run, it is sound to invoke
+            # `register_hook` on a known tensor.
```

Comment:
It's worth noting that you MUST NOT put register_hook calls in the graph for inputs. The canonical example is when the input tensor which had a registered hook is used differentiably outside of the context. The hook must be installed so that it triggers in that case too

Reply:
Yea, agreed :)